aggregator/operator_registrations_inmemory.go

```diff
@@ -0,0 +1,190 @@
+	operatorAddr common.Address
+	// channel through which to receive the response (operator pubkeys)
+	respC chan<- resp
+}
+type queryById struct {
+	operatorId types.OperatorId
+	// channel through which to receive the response (operator pubkeys)
+	respC chan<- resp
+}
+
+type resp struct {
+	operatorPubkeys types.OperatorPubkeys
+	// false if operators were not present in the pubkey dict
+	operatorExists bool
+}
+
+var _ operatorpubkeys.OperatorPubkeysService = (*OperatorRegistrationsServiceInMemory)(nil)
+
+// NewOperatorRegistrationsServiceInMemory constructs a OperatorRegistrationsServiceInMemory and starts it in a goroutine.
+// It takes a context as argument because the "backfilling" of the database is done inside this constructor,
+// so we wait for all past NewPubkeyRegistration events to be queried and the db to be filled before returning the service.
+// The constructor is thus following a RAII-like pattern, of initializing the serving during construction.
+// Using a separate initialize() function might lead to some users forgetting to call it and the service not behaving properly.
+func NewOperatorRegistrationsServiceInMemory(
+	ctx context.Context,
+	avsRegistrySubscriber avsregistry.AvsRegistrySubscriber,
+	avsRegistryReader avsregistry.AvsRegistryReader,
+	logger logging.Logger,
+) *OperatorRegistrationsServiceInMemory {
+	queryByAddrC := make(chan queryByAddr)
+	queryByIdC := make(chan queryById)
+
+	pkcs := &OperatorRegistrationsServiceInMemory{
+		avsRegistrySubscriber: avsRegistrySubscriber,
+		avsRegistryReader:     avsRegistryReader,
+		logger:                logger,
+		queryByAddrC:          queryByAddrC,
+		queryByIdC:            queryByIdC,
+	}
+
+	// We use this waitgroup to wait on the initialization of the inmemory pubkey dict,
+	// which requires querying the past events of the pubkey registration contract
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	pkcs.startServiceInGoroutine(ctx, queryByAddrC, queryByIdC, &wg)
+	wg.Wait()
+
+	return pkcs
+}
+
+func (ors *OperatorRegistrationsServiceInMemory) startServiceInGoroutine(ctx context.Context, queryByAddrC <-chan queryByAddr, queryByIdC <-chan queryById, wg *sync.WaitGroup) {
+	go func() {
+		pubkeyByAddrDict := make(map[common.Address]types.OperatorPubkeys)
+		pubkeyByIdDict := make(map[types.OperatorId]types.OperatorPubkeys)
+
+		ors.logger.Debug("Subscribing to new pubkey registration events on blsApkRegistry contract", "service", "OperatorRegistrationsServiceInMemory")
+		newPubkeyRegistrationC, newPubkeyRegistrationSub, err := ors.avsRegistrySubscriber.SubscribeToNewPubkeyRegistrations()
+		if err != nil {
+			ors.logger.Error("Fatal error opening websocket subscription for new pubkey registrations", "err", err, "service", "OperatorRegistrationsServiceInMemory")
+			// see the warning above the struct definition to understand why we panic here
+			panic(err)
+		}
+
+		ors.queryPastRegisteredOperatorEventsAndFillDb(ctx, pubkeyByAddrDict, pubkeyByIdDict)
```

Comment:
Could we lift this block outside the goroutine, leaving only the `for` in it? I think we could avoid the `WaitGroup` at all which makes things a bit confusing

Reply:
`WaitGroup::Wait` parks calling goroutine, allowing scheduler to execute another one until WaitGroup's counter is 0. Proposed implementation changes asynchronous code to synchronous where the thread will hang until `ors.queryPastRegisteredOperatorEventsAndFillDb` finishes and won't be able to do smth else.

Reply:
See `runtime_Semacquire`

Reply:
Seems mostly like a runtime difference on which thread the `ors.queryPastRegisteredOperatorEventsAndFillDb` will be executed. In both cases the caller of the constructor `NewOperatorRegistrationsServiceInMemory` will be blocked due to the `wg.Wait`